File: 347-top-k-frequent-elements/top-k-frequent-elements.go
func topKFrequent(nums []int, k int) []int {
    if len(nums) < 2 {
        return nums
    }

    counts := make(map[int]int)
    reverse_counts := make(map[int][]int)
    result := []int{}
    for _, num := range nums {
        counts[num]++
    }

    for num, count := range counts {
        reverse_counts[count] = append(reverse_counts[count], num)
    }

    i := len(nums)
    j := 0
    for len(result) < k {
        if list, ok := reverse_counts[i]; ok {
            if j < len(list) {
                result = append(result, list[j])
                j++
            } else {
                j=0
                i--
            }
        } else {
            i--
        }
    }
    return result
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
# given nums arr, k
# return k most frequent elements

# heap should be faster than sorting
# top k most frequent
    # so max heap, ordered by max frequency

import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequencies = dict()
        for i in nums:
            if i not in frequencies:
                frequencies[i] = 0
            frequencies[i] += 1
        topk = []
        for i in list(frequencies.keys()):
            topk.append((frequencies[i], i))
        
        final = []
        heapq.heapify_max(topk)
        for i in range(k):
            top = heapq.heappop_max(topk)
            final.append(top[1])
        return final